quinn-proto/src/connection/streams/mod.rs

```diff
@@ -335,60 +335,43 @@ impl<'a> SendStream<'a> {
     }
 }
 
-fn push_pending(pending: &mut BinaryHeap<PendingLevel>, id: StreamId, priority: i32) {
-    for level in pending.iter() {
-        if priority == level.priority {
-            level.queue.borrow_mut().push_back(id);
-            return;
-        }
-    }
-
-    // If there is only a single level and it's empty, repurpose it for the
-    // required priority
-    if pending.len() == 1 {
-        if let Some(mut first) = pending.peek_mut() {
-            let mut queue = first.queue.borrow_mut();
-            if queue.is_empty() {
-                queue.push_back(id);
-                drop(queue);
-                first.priority = priority;
-                return;
-            }
-        }
-    }
-
-    let mut queue = VecDeque::new();
-    queue.push_back(id);
-    pending.push(PendingLevel {
-        queue: RefCell::new(queue),
-        priority,
-    });
+/// Push a pending stream ID with the given priority, queued after all preexisting streams for the priority
+fn push_pending(pending: &mut BTreeMap<PendingPriority, StreamId>, id: StreamId, priority: i32) {
+    // Get all preexisting streams of this priority
+    let mut range = pending.range((
+        Bound::Included(PendingPriority {
+            priority,
+            recency: 0,
+        }),
+        Bound::Included(PendingPriority {
+            priority,
+            recency: u64::MAX,
+        }),
+    ));
+
+    // Determine the recency value for this stream
+    // Setting it to 1 below all preexisting streams of this priority causes it to be sorted after all of those streams in
+    // the `BTreeMap`. This implements round-robin scheduling for streams that are still pending even after being handled,
+    // as in that case they are removed and then immediately reinserted.
+    // If this is the only/first stream for this priority, recency is initialised to u64::MAX instead. As this function is
+    // the only place that recency is initialised or mutated, this ensures that it will practically never underflow.
+    let recency = range.next().map(|(p, _)| p.recency - 1).unwrap_or(u64::MAX);
```

Comment:
Could we simplify this by using a single `Connection`-scoped recency value that we decrement on each insert? `u64` is big enough that this still won't ever overflow, and it'd save a traversal.

Also might as well use wrapping arithmetic to update it, since a rollover would be harmless even if it did somehow happen.

Reply:
> Could we simplify this by using a single Connection-scoped recency value that we decrement on each insert? u64 is big enough that this still won't ever overflow, and it'd save a traversal.

This is an excellent suggestion, and actually lets us switch back to using `BinaryHeap`, as needing to do lookups on the collection was the whole motivation for replacing it with a `BTreeMap`, even though `BinaryHeap` is a better fit for the other ways the collection is used.

> Also might as well use wrapping arithmetic to update it, since a rollover would be harmless even if it did somehow happen.

An underflow would break fairness, as the recency of the stream that caused the underflow would wrap around to `u64::MAX`, while all other streams of that priority would still have recency values close to 0, so only that one stream would be handled until its recency gets close to 0 again.

Reply:
Oh, right, we might never touch the pre-underflow streams again. Good thing `u64` is big enough that this won't happen in practice.